crates/lib/src/qvm/api.rs

```diff
@@ -0,0 +1,357 @@
+    Multishot,
+    MultishotMeasure,
+    Expectation,
+    Wavefunction,
+}
+
+#[derive(Debug, Deserialize, Clone, Eq, PartialEq)]
+pub(super) struct Failure {
+    /// The message from QVM describing what went wrong.
+    pub(super) status: String,
+}
+
+/// A QVM response that can be deserialized into some successful response type `T`, or into a
+/// [`Failure`] containing an error message.
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+#[serde(untagged)]
+pub(super) enum QvmResponse<T>
+where
+    T: DeserializeOwned,
+{
+    #[serde(bound = "")]
+    Success(T),
+    Failure(Failure),
+}
+
+/// Fetch the version information from the running QVM server.
+pub async fn get_version_info(config: &ClientConfiguration) -> Result<String, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!("requesting qvm version information");
+    let client = reqwest::Client::new();
+    let params = vec![("type", "version")];
+    client
+        .post(config.qvm_url())
+        .json(&params)
+        .send()
+        .await
+        .map_err(|source| Error::QvmCommunication {
+            qvm_url: config.qvm_url().into(),
+            source,
+        })?
+        .text()
+        .await
+        .map_err(|source| Error::QvmCommunication {
+            qvm_url: config.qvm_url().into(),
+            source,
+        })
+}
+
+/// Executes a program on the QVM.
+pub async fn run(
+    request: &MultishotRequest,
+    config: &ClientConfiguration,
+) -> Result<MultishotResponse, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!("making a multishot request to the QVM");
+    let response = make_request(request, config).await?;
+    match response.json::<QvmResponse<MultishotResponse>>().await {
+        Ok(QvmResponse::Success(response)) => Ok(response),
+        Ok(QvmResponse::Failure(response)) => Err(Error::Qvm {
+            message: response.status,
+        }),
+        Err(source) => Err(Error::QvmCommunication {
+            qvm_url: config.qvm_url().into(),
+            source,
+        }),
+    }
+}
+
+/// The request body needed to make a multishot [`run`] request to the QVM.
+#[derive(Serialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+pub struct MultishotRequest {
+    /// The Quil program to run.
+    pub quil_instructions: String,
+    /// The memory regions to include in the response.
+    pub addresses: HashMap<String, AddressRequest>,
+    /// The number of trials ("shots") to run.
+    pub trials: u16,
+    #[serde(rename = "type")]
+    request_type: RequestType,
+}
+
+/// An enum encapsulating the different ways to request data back from the QVM for an address.
+#[derive(Serialize, Debug, Clone, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum AddressRequest {
+    /// Set to `true` to request all values at this address, or `false` to explicitly exclude
+    /// the address.
+    All(bool),
+    /// A list of specific indices to get back for the address.
+    Indices(Vec<usize>),
+}
```

Comment:
Still workshopping the name of this enum and it's variants, suggestions welcome.

Reply:
For more context, the QVM allows you to specify what memory regions you want back in the response with a map like:

```
{
    "ro": true,       # returns all values for "ro" 
    "foo": false,     # omit "foo" entirely
    "bar": [3, 5]     # return the values in bar[3] and bar[5]
}
```